### ISSUE TITLE
Use CMAKE_IOS_INSTALL_COMBINED CMake Variable for Universal Library Support Out-of-the-Box

### ISSUE THREAD

**Issue by epicstar** [leetal/ios-cmake#12]
Since CMake 3.5.2, it seems CMake has support for Universal fat libraries out-of-the-box. This can be set with [`CMAKE_IOS_INSTALL_COMBINED`](https://cmake.org/cmake/help/v3.5/variable/CMAKE_IOS_INSTALL_COMBINED.html).

From the official documentation for the target property [`IOS_INSTALL_COMBINED`](https://cmake.org/cmake/help/v3.5/prop_tgt/IOS_INSTALL_COMBINED.html#prop_tgt:IOS_INSTALL_COMBINED):

> When this property is set to set to false (which is the default) then it will either be built with the device SDK or the simulator SDK depending on the SDK set. But if this property is set to true then the target will at install time also be built for the corresponding SDK and combined into one library.

> This feature requires at least Xcode version 6.

Internally, it seems to create an install task on the target to build for all the supplied Apple Device SDKs then combine all the binaries into 1 fat binary using `lipo`.

I haven't actually started using this variable yet,  and I'm not exactly sure how it would even fit with the current toolchain, but if I can figure it out before someone else investigates it, I'm willing to create a pull request.

**Comment by epicstar** [comment]
For those interested, here is the [CMakeIOSInstallCombined](https://github.com/Kitware/CMake/blob/master/Modules/CMakeIOSInstallCombined.cmake) module source code.

**Comment by leetal** [comment]
Great find! :D
Looking at the CMake tests it actually looks like this has in some form worked since CMake 3.3+ even. At least during development releases.

**Comment by epicstar** [comment]
Ok, so I ended up getting my build to run googletests on iOS XCode (complicated, if you're interested, PM me), but as expected, while my basic build and test targets work as expected, I can't get the install to work correctly. I need to figure it out but it seems that the standard CMake target install with `CMAKE_IOS_INSTALL_COMBINED` enabled to true with the toolchain conflict with each other. I'll try to see what I can do to fix it.

**Comment by epicstar** [comment]
@leetal, I actually got this to work. I tested this by using `-DIOS_PLATFORM=OS -DIOS_DEPLOYMENT_TARGET=8.0`:

I got this working by doing the following:

## 1. set [`CMAKE_OSX_DEVELOPMENT_TARGET`](https://cmake.org/cmake/help/v3.11/variable/CMAKE_OSX_DEPLOYMENT_TARGET.html) to match `IOS_DEPLOYMENT_TARGET`

* `CMAKE_OSX_DEVELOPMENT_TARGET` isn't set in the toolchain for some reason (prolly because you're using a CMake version older than 3.11 at your workplace), so whenever I ran the `install` cmake target, cmake would force a rebuild and set the `-mios-${platform}-version-min` to whatever the default iOS version is and ignore the one set in 
* Note that you no longer need to set that flag I was talking about anymore if using CMake 3.11 since this is automatically added by CMake via this CMake variable


## 2. You must add the simulator  architectures as `XCODE_ATTRIBUTES` (see below) into your `CMakeLists.txt`:  

```  # when using the install target, make sure everything builds as expected.
if (CMAKE_IOS_INSTALL_COMBINED AND CMAKE_OSX_SYSROOT MATCHES "os")
  set_target_properties(fooTarget bazTarget PROPERTIES
      XCODE_ATTRIBUTE_ARCHS[sdk=iphonesimulator*] i386 x86_64
      XCODE_ATTRIBUTE_VALID_ARCHS[sdk=iphonesimulator*] i386 x86_64
  )
endif()
```

 * This ensures that CMake's install target will pick up the correct architectures when running the `ios_install_combined` when running the `install` target
* `ios_install_combined` looks for both properties on the target
* if you start with `-DIOS_PLATFORM=SIMULATOR[64]`, you will have to set the opposite on the targets. (but this will only build for `x86_64`, you can't build `i686` 32-bit with the toolchain, and I decided not to investigate this angle)
* If setting the architectures on the same platform, I get a CRAPTON of linker errors if creating an executable from a library
* I don't fully understand why this works, but it does...

-----------------------------------

But yeah, doing these two things fixes the problems. I found these two out by looking at the CMake repo, and most of the changes on the iOS build code of CMake come from @ruslo. 

I believe (1) is easy to solve by adding 1 simple line to the toolchain (just add flag if cmake is 3.11 or higher, else fall back to other behavior). However, (2) will probably only work by changing the documentation since this is dependent on users setting the correct architectures on their targets.

This begs a bigger question if CMake should just have 1st party support for iOS like it has for Android. I believe your toolchain is definitely good enough especially if the code is cleaned for the newer versions of CMake. CMake really should clean up their documentation for modern practices, also.

**Comment by ruslo** [comment]
> I found these two out by looking at the CMake repo, and most of the changes on the iOS build code of CMake come from @ruslo.

The state of this feature is a little bit messy because in fact iOS platform is still not officially supported by CMake :) As far as I know there is some progress there, see this pull request: https://gitlab.kitware.com/cmake/cmake/issues/17870

In practice this feature will not work for random iOS toolchain and require some agreements (e.g. it will work only for Xcode and `CMAKE_OSX_SYSROOT` should be set to `iphoneos`), see hints in tests:
* https://github.com/Kitware/CMake/blob/v3.12.2/Tests/RunCMake/XcodeProject/XcodeIOSInstallCombinedPrune.cmake
* https://github.com/Kitware/CMake/blob/v3.12.2/Tests/RunCMake/XcodeProject/XcodeIOSInstallCombined.cmake
* https://github.com/Kitware/CMake/blob/v3.12.2/Tests/RunCMake/XcodeProject/XcodeIOSInstallCombinedSingleArch.cmake

You can find the real-world toolchains in this collection: https://github.com/ruslo/polly

For example:
* https://github.com/ruslo/polly/blob/master/ios-11-4-dep-9-3.cmake
* https://github.com/ruslo/polly/blob/master/os/iphone.cmake

**Comment by patrikhuber** [comment]
Hi! This question may fit well here. I see in @ruslo's link that the official cmake iOS support is coming along well, but it might not just be 100% there yet.

I'd like to ask, what are the main differences between @ruslo's polly iOS toolchain files, and @leetal's iOS toolchain files in this repository? Both look quite mature and well-working to me. Any hints on what the differences might be and when to best go for one or the other?

**Comment by ruslo** [comment]
> I see in @ruslo's link that the official cmake iOS support is coming along well, but it might not just be 100% there yet

I haven't tested it yet, I guess I will need to tweak few variables in Polly to match CMake 3.14 (?) iOS support. In general as far as I understand [Polly](https://github.com/ruslo/polly), [CMake iOS support](https://gitlab.kitware.com/cmake/cmake/issues/17870) and https://github.com/leetal/ios-cmake are independent things.

> the main differences between @ruslo's polly iOS toolchain files, and @leetal's iOS toolchain files in this repository?

According to the README:
- https://github.com/leetal/ios-cmake#options

```
The resulting binary will consist of only one platform. To combine all
platforms into the same fat-library, use the LIPO tool. More information
on this is available on the net
```

It's not designed to use both simulator and device simultaneously, if you want such library you have to independently build simulator, then build device, then use lipo.

With Polly, simulator and device architectures can be mixed together without any extra steps (see [CMAKE_IOS_INSTALL_COMBINED](https://cmake.org/cmake/help/v3.5/variable/CMAKE_IOS_INSTALL_COMBINED.html) feature). E.g. build with `ios-nocodesign-11-4-dep-9-3` toolchain:
- https://travis-ci.org/ingenue/hunter/jobs/497488024

Will produce the 5 architectures library:
```
> mkdir __temp && cd __temp
> wget https://github.com/ingenue/hunter-cache/releases/download/cache-19005d5/19005d5e7164d248ee51ba10ac03caf6d0d48645.tar.bz2
> tar xf 19005d5e7164d248ee51ba10ac03caf6d0d48645.tar.bz2
> lipo -info lib/libicui18n.a
Architectures in the fat file: lib/libicui18n.a are: i386 armv7 armv7s x86_64 arm64
```

This library produced by regular `cmake --build _builds --target install` command and package can be injected to user code with regular `find_package(... CONFIG REQUIRED)`.

**Comment by patrikhuber** [comment]
I see! So I would deduce from that that leetal/ios-cmake is quite a basic (while well-working) iOS toolchain file for CMake, while polly's iOS toolchain files offer many more options (like e.g. the mixing of architectures that you mentioned).

Thank you very much, that's a great answer.

**Comment by leetal** [comment]
Fixed in [v3.0.0](https://github.com/leetal/ios-cmake/tree/3.0.0) 🔥 See #27 for more information to official docs.
